### PR TITLE
Prevent linking to TNS When an Object ID is located in a URL

### LINKED ISSUE #39: Auto-Linking Conflict between hyperlinks and Transient Name Server objects
# Current behavior
I am unsure of the prevalence of this issue, but [Circular 31683](https://gcn.nasa.gov/circulars/31683) demonstrates an issue where the automatic hyperlinking system conflicts with the system that automatically recognizes and links to items in the Transient Name Server [(TNS)](https://www.wis-tns.org/). 

In this Circular, there is a hyperlink to an image located offsite, but the URL  includes a TNS Object ID, which results in the hyperlink being cut short and not linking correctly due to the system automatically inserting a link to the TNS page. 

# Expected behavior
TNS Objects should be recognized and linked to when referenced in text, but linking to an in-text URL should take precedence over linking to TNS

# Steps to reproduce

1. Navigate to [Circular 31683](https://gcn.nasa.gov/circulars/31683)
2. Following the second paragraph, the Circular links to a figure located offsite. The URL is http://grb.rssi.ru/GRB220219B/GRB220219B_AT2022cva_LC.png
3.  Because the URL contains the TNS Object ID "AT2022cva," the system automatically converts the "AT2022cva" into `[AT2022cva](https://www.wis-tns.org/object/2022cva)` and the URL then becomes two hyperlinks: [http://grb.rssi.ru/GRB220219B/GRB220219B_](http://grb.rssi.ru/GRB220219B/GRB220219B_) [AT2022cva](https://www.wis-tns.org/object/2022cva)_LC.png . This breaks the intended hyperlink.

# Environment
Bug discovered on Windows 11 Chrome, but likely not system specific

### PR DESCRIPTION
**Pull request by tylerbarna**
# Description
This modifies the regular expression that looks for TNS Object IDs to exclude URL strings. This prevents the automatic linking of TNS Objects from interfering with hyperlinks. This is accomplished by using negative lookbehind terms in the regular expression that exclude any strings that start with 'https://' or 'http://'

# Related Issue(s)

Resolves #39 

# Testing
On the GCN website, [Circular 31683](https://gcn.nasa.gov/circulars/31683) documents the described issue just after the 2nd paragraph, where a link to a file is broken due to the URL containing a TNS ID:
image

I updated the regular expression used to find TNS Objects in a forked version of this repository and then installed this fork to my local GCN development environment. I added Circular 31683 to the [sandbox-seed.json](https://github.com/tylerbarna/gcn.nasa.gov/blob/d7dbcaf42198769a5efac80121df219cd60c5b33/sandbox-seed.json) file in order to test the issue.

After navigating to Circular 31683 in the local development environment, one can see that the URL hyperlink is now rendered correctly, and other references to AT2022cva are unaffected:

image

I have also attempted to add a test case to the testing files in this repo, but I would appreciate a review of this aspect to ensure I have implemented it correctly

### CODE REVIEW

File: src/replacements/tns/index.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import type { FindAndReplace } from '../index.js'
 
 export default {
-  find: /AT\s*(2\d{3}[a-z]+)/g,
+  find: /(?<!http:\/\/|https:\/\/)(?<!http:\/\/.*|https:\/\/.*)AT\s*(2\d{3}[a-z]+)/g,
```
Comment (dakota002) [inline src/replacements/tns/index.ts:4]: You may be able to simplify the regex:
```suggestion
  find: /(?<!http(s?):\/\/)(?<!http(s?):\/\/.*)/g,
```
the `s?` should match 0 or 1 of the character `s`, can you try this and see if it works?

Comment (tylerbarna) [reply src/replacements/tns/index.ts:4]: that seems to work, thanks!

**Review by lpsinger — COMMENTED** [review]
I think that this will fail to correctly format a paragraph that contains a valid URL _followed by_ a TNS name, as in:

```https://google.com/ is a URL but AT2017gfo is not```

**Comment by tylerbarna** [comment]
> I think that this will fail to correctly format a paragraph that contains a valid URL _followed by_ a TNS name, as in:
> 
> `https://google.com/ is a URL but AT2017gfo is not`

Just verified this is the case, will fix

**Comment by tylerbarna** [comment]
looking into why the checks are failing

File: src/replacements/tns/index.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import type { FindAndReplace } from '../index.js'
 
 export default {
-  find: /AT\s*(2\d{3}[a-z]+)/g,
+  find: /(?<!https?:\/\/)(?<!https?:\/\/[^ ]*)AT\s*(2\d{3}[a-z]+)/g,
```
Comment (lpsinger) [inline src/replacements/tns/index.ts:4]:
```suggestion
  find: /(?<!https?:\/\/\S*)AT\s*(2\d{3}[a-z]+)/g,
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Review by lpsinger — APPROVED** [review]

**Pull request merged**